rclcpp_lifecycle/include/rclcpp_lifecycle/lifecycle_node.hpp

```diff
@@ -1056,6 +1056,11 @@ class LifecycleNode : public node_interfaces::LifecycleNodeInterface,
   bool
   register_on_error(std::function<LifecycleNodeInterface::CallbackReturn(const State &)> fcn);
 
+  RCLCPP_LIFECYCLE_PUBLIC
+  void
+  send_callback_resp(
```

Comment:
Why is this a public API? This seems problematic and error-prone to allow the user to send a callback response at any point.

Reply:
Circling back I think I understand why you implemented this as a public API, in order for the user to call it and send the DEFER callback return without needing a new function or overload for the transition callbacks, yes? What happens when the user calls this function outside of a transition? Besides potential misuse or errors, I don't know if this as a public API really communicates how and when it is actually supposed to be used.

Reply:
Yes, that is largely the reasoning with the other reason being it is a bit more consistent to have the user use the `lifecycle_node.hpp` public API (although [prior implementation](https://github.com/tgroechel/rclcpp/blob/change_state_handler_obj/rclcpp_lifecycle/include/rclcpp_lifecycle/change_state_handler.hpp) has a public API class with a single function so not as strong of an argument). Currently, calling `send_callback_resp` checks the state of the node and proceeds accordingly, returning a `RCL_RET_ERROR` if called when the node in not in a transitionary state.

I agree, I think you lose some explicit use with the benefit of less duplicated (and easier to understand/maintain in my opinion) backend code. However, the passing of a pointer to a `ChangeStateHandler` has a similar concern, just more mitigated (what happens when someone calls `send_callback_resp` multiple times from the `ChangeStateHandlerPtr`?).

With either approach, I think there needs to be a better mechanism for preventing calling `send_callback_resp` improperly. I will think more about it but initial thoughts are around:
1. the user needing to specify where they are calling it from (e.g., `send_deferred_callback_resp(ON_CONFIGURE, SUCCESS)`) 
2. some token system (i.e., give a token that can only be used once). 

The former would be easier to implement / keep prior LC structure whereas the latter would likely need a change to the callback function definitions (so back to the parallel map or changing the default function def).

Reply:
https://github.com/tgroechel/rclcpp/commit/89c9fc1f1880a5d0e5c6ea9588df188fb3b721f6